smartsim/settings/dispatch.py

```diff
@@ -0,0 +1,290 @@
+        # Disabling because we want to match the the type of the dispatch
+        # *exactly* as specified by the user
+        # pylint: disable-next=unidiomatic-typecheck
+        return type(launcher) is self.launcher_type
+
+    def create_new_launcher_configuration(
+        self, for_experiment: Experiment, with_settings: _DispatchableT
+    ) -> _LaunchConfigType:
+        launcher = self.launcher_type.create(for_experiment)
+        return self.create_adapter_from_launcher(launcher, with_settings)
+
+    def create_adapter_from_launcher(
+        self, launcher: LauncherProtocol[_LaunchableT], settings: _DispatchableT
+    ) -> _LaunchConfigType:
+        if not self._is_compatible_launcher(launcher):
+            raise TypeError(
+                f"Cannot create launcher adapter from launcher `{launcher}` "
+                f"of type `{type(launcher)}`; expected launcher of type "
+                f"exactly `{self.launcher_type}`"
+            )
+
+        def format_(exe: ExecutableProtocol, env: _EnvironMappingType) -> _LaunchableT:
+            return self.formatter(settings, exe, env)
+
+        return _LauncherAdapter(launcher, format_)
+
+    def configure_first_compatible_launcher(
+        self,
+        with_settings: _DispatchableT,
+        from_available_launchers: t.Iterable[LauncherProtocol[t.Any]],
+    ) -> _LaunchConfigType:
+        launcher = helpers.first(self._is_compatible_launcher, from_available_launchers)
+        if launcher is None:
+            raise errors.LauncherNotFoundError(
+                f"No launcher of exactly type `{self.launcher_type.__name__}` "
+                "could be found from provided launchers"
+            )
+        return self.create_adapter_from_launcher(launcher, with_settings)
+
+
+@t.final
+class _LauncherAdapter(t.Generic[Unpack[_Ts]]):
+    def __init__(
+        self,
+        launcher: LauncherProtocol[_LaunchableT],
+        map_: t.Callable[[Unpack[_Ts]], _LaunchableT],
+    ) -> None:
+        # NOTE: We need to cast off the `_LaunchableT` -> `Any` in the
+        #       `__init__` method signature to hide the transform from users of
+        #       this class. If possible, this type should not be exposed to
+        #       users of this class!
+        self._adapt: t.Callable[[Unpack[_Ts]], t.Any] = map_
+        self._adapted_launcher: LauncherProtocol[t.Any] = launcher
+
+    def start(self, *args: Unpack[_Ts]) -> LaunchedJobID:
+        payload = self._adapt(*args)
+        return self._adapted_launcher.start(payload)
+
+
+DEFAULT_DISPATCHER: t.Final = Dispatcher()
+# Disabling because we want this to look and feel like a top level function,
+# but don't want to have a second copy of the nasty overloads
+# pylint: disable-next=invalid-name
+dispatch: t.Final = DEFAULT_DISPATCHER.dispatch
+
+
+# >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
+# TODO: move these to a common module under `smartsim._core.launcher`
```

Comment:
Should we do this before merging PR?

Reply:
Hahaha good question! I'm willing to do it if others aren't too worried about the size of the PR

Reply:
Decided to move this work into a follow on ticket